docs/topics/exceptions.rst

```diff
@@ -14,13 +14,14 @@ Built-in Exceptions reference
 
 Here's a list of all exceptions included in Scrapy and their usage.
 
-DropItem
---------
+CancelRequest
+-------------
 
-.. exception:: DropItem
+.. exception:: CancelRequest
 
-The exception that must be raised by item pipeline stages to stop processing an
-Item. For more information see :ref:`topics-item-pipeline`.
+Raised to indicate that no further bytes should be downloaded for a request.
```

Comment:
I think the approach of raising an exception from a singnal handler is good, but we need to explain better the semantics, and check how it works with other parts of Scrapy.

1. Signal name. For cancelled request I'd not expect callback to be called. Here we're not cancelling a request (request is already sent); we're stopping/aborting download of the response. 
2. Scrapy can also abort download on different conditions, such as max response size. I think in this case errback is called, not callback. Should we mention this in docs? There is also DOWNLOAD_FAIL_ON_DATALOSS option which look related, though I'm not sure it should affect the behavior of the cancel exception. 
3. I'm not sure what's better, call a callback or an errback, and if there is a single best approach. Maybe there should be 2 exceptions (or a bool flag in exception), one which make download successful, and another one for errors. E.g. in the FAQ entry you're writing "detect early that its response won't be useful", but if response is not useful, it shouldn't go to a callback.
4. Are there other places where this exception can be raised? I don't have anything in mind, but the CancelRequest exception look similar to IgnoreRequest which can be raised by a downloader middlewares; this can be confusing. Are there use cases for catching CancelRequest from other signals?

Reply:
Those are very good points, thank you.

(1) I don't particularly love the current name, and I agree with the fact that the "request is already sent", so it's not completely accurate. What about `(Abort/Interrupt/Stop)Response`?

(2/3) I think that since this behaviour would be controlled explicitly by user code, and not because of an unexpected error, we should call the assigned callback. That said, providing a way to fire the errback instead sounds fine to me too. What about an `errback` boolean argument in the exception, which defaults to `False`?

(4) I don't think we could catch this exception anywhere else, since the signal is directly sent from the download handler. Responses in middlewares are already complete.

Reply:
(1) Based on the documentation, `(Abort/Cancel/Interrupt/Stop)Download`?

Reply:
Sounds even better. At point point I thought about `(Verb)(Response/Body)Download`, since what we interrupt is technically only the body, but that was too verbose, and probably unnecessary.

Reply:
I went for `StopDownload`. Also reworded the docs from "the response might not be useful" to "the already downloaded content might be enough". Given that (and my comment below about this being "controlled explicitly by user code") I think it'd be good to fire the callback instead of the errback. Thoughts?

Reply:
Adding a `DOWNLOAD_FAIL_ON_DOWNLOAD_STOP` (or `DOWNLOAD_FAIL_ON_STOP`?) setting may make sense, like `DOWNLOAD_FAIL_ON_DATALOSS`.

I have no strong on what the default behavior should be. While the error is caused by users, as @elacuesta said, I can imagine users desiring both behaviors:
- If the signal is used to stop downloads once the desired information has been downloaded, `callback` makes sense.
- If the signal is used to stop downloads as soon as they are detected as bad responses based on content, `errback` makes sense to cause a retry (as it happens for data loss with `DOWNLOAD_FAIL_ON_DATALOSS=True`).